### PR TITLE
std-list-view: bring selected into view

### LINKED ISSUE #2209: Selecting a new item in StandardListView by keyboard does not trigger scrolling
Items inside of a StandardListView can be selected by keyboard arrow keys, but doing so will not scroll if the newly selected item is outside the current viewport. This can be easily tested in the widget gallery by selecting an item and then using down arrow (several times) to change the selection to an item in the bottom of the list.

### PR DESCRIPTION
**Pull request by FloVanGH**
* Bring selected item on `StandardListView` into view by arrow keys and click
* works for native, material and fluent

Fixes #2209

### CODE REVIEW

File: internal/compiler/widgets/fluent-base/std-widgets-base.slint [hunk 1]
```diff
@@ -519,19 +523,38 @@ component StandardListViewBase inherits ListView {
...
+        }
+
+        root.current-item = index;
+
+        if(current-item-y < 0) {
+            self.viewport-y += 0 - current-item-y;
+        }
+        
+        if(current-item-y + item-height > self.visible-height) {
+            self.viewport-y -= current-item-y + item-height - self.visible-height;
```
Comment (tronical) [inline internal/compiler/widgets/fluent-base/std-widgets-base.slint:545]: I think you can omit `root` and `self` qualifiers here. Another option would be to consistently use `self`. A matter of taste I guess :-)

```suggestion
        current-item = index;

        if(current-item-y < 0) {
            viewport-y += 0 - current-item-y;
        }
        
        if(current-item-y + item-height > visible-height) {
            viewport-y -= current-item-y + item-height - visible-height;
```

**Review by tronical — COMMENTED** [review]
Nice! :) I think it would be good to also document the new `set-current-item` function (worth a ChangeLog entry as well maybe).

**Comment by medhefgo** [comment]
So anyone using a Flickable is out of the picture? :(

**Comment by FloVanGH** [comment]
@medhefgo 

> So anyone using a Flickable is out of the picture? :(

First of all thank you for your feedback. 

To your note: I think the behavior implemented with this PR is a good extension for the `StandardListView`. The implementation of a generic function like  `bring_into_view` functionality for `Flickable` that also works with items of different height is much more complex. That also the reason why stuff like virtualization of list items is also difficult on lists with different item heights and therefore mostly only implemented for list with items with the same height. For now for custom components using a  `Flickable` I think it's already possible to implement your custom logic, because you can access e.g. `viewport-y`. But I agree it would be great to have something generic built-in  `Flickable` in the future. Probably it will be necessary to write native code to archive that behavior.

**Comment by ogoffart** [comment]
> So anyone using a Flickable is out of the picture? :(

For now, anyone with a Flickable will have to do it manually.
This is not too hard, and we should probably have a recipes in our documentation for it:

```
                    if (-fli.viewport-x > item.x) {
                        fli.viewport-x = -item.x;
                    }
                    if (-fli.viewport-y > item.y) {
                        fli.viewport-y = -item.y;
                    }
                    if (-fli.viewport-x + fli.width < item.x + item.width) {
                        fli.viewport-x = -item.x + fli.width - item.width;
                    }
                    if (-fli.viewport-y + fli.height < item.y + item.height) {
                        fli.viewport-y = -item.y + fli.height - item.height;
                    }
```

[example](https://slint-ui.com/releases/0.3.4/editor/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A++++height%3A+300px%3B%0A++++width%3A+300px%3B%0A++++VerticalBox+%7B%0A++++++++fli+%3A%3D+Flickable+%7B%0A++++++++++++viewport-height%3A+2500px%3B%0A++++++++++++viewport-width%3A+2500px%3B%0A++++++++++++Rectangle+%7B+%0A++++++++++++++++background%3A+%40linear-gradient%2875deg%2C+blue%2C+orange%2C+yellow%2C+green%29%3B%0A++++++++++++%7D%0A++++++++++++item+%3A%3D+Rectangle+%7B+%0A++++++++++++++++background%3A+red%3B%0A++++++++++++++++x%3A+500px%3B%0A++++++++++++++++y%3A+860px%3B%0A++++++++++++++++width%3A+30px%3B%0A++++++++++++++++height%3A+25px%3B%0A++++++++++++%7D%0A%0A++++++++%7D%0A++++++++HorizontalLayout+%7B+%0A++++++++++++alignment%3A+center%3B%0A++++++++++++Button+%7B+%0A++++++++++++++++text%3A+%22Make+Visible%22%3B+%0A++++++++++++++++clicked+%3D%3E+%7B%0A++++++++++++++++++++if+%28-fli.viewport-x+%3E+item.x%29+%7B%0A++++++++++++++++++++++++fli.viewport-x+%3D+-item.x%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++++++if+%28-fli.viewport-y+%3E+item.y%29+%7B%0A++++++++++++++++++++++++fli.viewport-y+%3D+-item.y%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++++++if+%28-fli.viewport-x+%2B+fli.width+%3C+item.x+%2B+item.width%29+%7B%0A++++++++++++++++++++++++fli.viewport-x+%3D+-item.x+%2B+fli.width+-+item.width%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++++++if+%28-fli.viewport-y+%2B+fli.height+%3C+item.y+%2B+item.height%29+%7B%0A++++++++++++++++++++++++fli.viewport-y+%3D+-item.y+%2B+fli.height+-+item.height%3B%0A++++++++++++++++++++%7D%0A++++++++++++++++%7D%0A++++++++++++%7D+%0A++++++++%7D%0A++++%7D%0A%7D%0A)


But yes, we might want to add an `ensure-visible()` function or something in the future

**Review by ogoffart — APPROVED** [review]

**Pull request merged**